### ISSUE TITLE
Implementation of Runtime Interceptors for Enhanced Development in .NET

### ISSUE THREAD

**Issue by dotnetian** [dotnet/runtime#97453]
Just-in-time compilation (JIT) has been the traditional method for executing .NET programs for a long time, but it comes with certain limitations. This issue suggests using a concept similar to runtime interceptors, as seen in some languages like Python, in .NET development to improve the development experience, especially minimizing the time needed for build processes.

# Challenges with JIT, and Native AOT

The main purpose of the JIT compilation method is to make a balance between build times and runtime performance. However, being the best in none of these was possible by JIT compilation in .NET; Because some other languages were focusing only on a few performance factors, so they would be better in those focused fields.

And about .NET, Native AOT compilation was introduced in .NET 8, to focus on overall runtime performance and attempt to maximize it. So we can say that in Native AOT, the focus on the ease of development cycle was thrown away, which made Native AOT way better in production, and of course, developers still have JIT for development builds.

And this is the main reason I'm writing this issue. JIT has been there for .NET for a long time, and there wasn't a Native AOT compilation back then. Now that we have a build method that is best optimized for production, why not have another one optimized for development? JIT build time is not so slow, but especially when the solutions scale in the number of projects, dependencies, and functionalities, the build time noticeably drops.

Currently, the main thing that a rapid development process is focusing on is minimizing the time being spent for each build iteration, and some other factors like memory usage or execution performance are not in priority in development. And JIT simply is not the best in build speed. That's why I suggest implementing another building method for the .NET ecosystem that can fantastically solve the build time problem.

# Introducing the Concept of Runtime Interceptors

Runtime interceptors are powerful tools used in some languages to execute code immediately at runtime. So a separate tool, called an interceptor, reads the code line by line and provides the functionality for the programs. By incorporating a similar concept into the .NET ecosystem, developers can achieve faster feedback loops and reduce downtime during development. Another benefit of using interceptors is greatly improving Unit Tests' runtime, so a test that reports took 400ms will not take so many seconds to just build the projects.

# Utilizing Runtime Interceptors in .NET

To leverage runtime interceptors in .NET, tooling needs to be developed specifically for the .NET ecosystem. This program would provide the necessary infrastructure to resolve dependencies and execute code immediately at runtime, similar to how interceptors function in other languages.

# Conclusion

.NET languages, especially C#, currently do look like utopian programming languages a lot. They have clean and high-potential code syntax, remarkable runtime performance thanks to Native AOT, and a mixture of a billion-dollar company and a big community's support, all trying to make .NET better. Thanks to Native AOT and JIT, .NET has a great performance in runtime and outperforms many popular languages, but in development, it's not crucial, but if .NET utilizes runtime interceptors for maximizing development experience, it will be fantastic.

**Comment by huoyaoyuan** [comment]
> The main purpose of the JIT compilation method is to make a balance between build times and runtime performance. However, being the best in none of these was possible by JIT compilation in .NET; Because some other languages were focusing only on a few performance factors, so they would be better in those focused fields.
> 
> And about .NET, Native AOT compilation was introduced in .NET 8, to focus on overall runtime performance and attempt to maximize it. So we can say that in Native AOT, the focus on the ease of development cycle was thrown away, which made Native AOT way better in production, and of course, developers still have JIT for development builds.

This is not correct. The major advantage of JIT compilation should be portability. Most assemblies are just deployed with AnyCPU.
Also note that **JIT provides better steady-state performance than AOT**. With tiered compilation, PGO and other post-first run optimizations, JIT can optimize the program based on its actual runtime behavior. For example, the JIT can optimize static readonly fields into constants because their values are already assigned in tiered code.

NativeAOT was introduced in .NET 7. It focuses more on *startup* performance and conforming environment limitations. It does have opportunity to improve performance of some cases by knowing that *no more code will be loaded afterwards*.

> Currently, the main thing that a rapid development process is focusing on is minimizing the time being spent for each build iteration

R2R compilation is used in maximized performance configuration, which also requires a lot of built time.

> So a separate tool, called an interceptor, reads the code line by line and provides the functionality for the programs.

What about C# interactive and F# interactive?

> Another benefit of using interceptors is greatly improving Unit Tests' runtime, so a test that reports took 400ms will not take so many seconds to just build the projects.

Is it to avoid building other code in the same unit test project?
Many unit test projects use shared helpers to simplify individual unit tests. To figure out the helpers required, the majority process of building would be required.

**Comment by dotnetian** [comment]
Thank you so much @huoyaoyuan for fixing the problems. I updated the issue's text.

> NativeAOT was introduced in .NET 7. It focuses more on startup performance and conforming environment limitations. It does have opportunity to improve performance of some cases by knowing that no more code will be loaded afterwards.

You're right. I know that Native AOT is not as well-performing as JIT for now, but it's not so long since it has been released and I believe that it has a higher potential to have improvements, compared to a compilation method that .NET is using for decades. Please note that I don't say "JIT is bad"; JIT is super-stable and has been there for so long, so a lot of its problems have been solved. Also, I understand that because of the way that JIT works, it can do some performance optimizations that Native AOT simply can't.

> R2R compilation is used in maximized performance configuration, which also requires a lot of built time.

That's right too. But what I said was "a rapid _development_ process", I was not discussing production, and in development, startup time being 14ms or 800ms does not make such a difference, or loading an ASP.NET Core page in 40ms or 120ms is not even easily noticeable. These factors are critical in production, where tons of requests are being sent at the same time. I hope this example helps.

> What about C# interactive and F# interactive?

Great idea! language interactives and REPL tools that .NET have have done a lot of stuff needed to make a working interceptor for .NET Runtime. They can resolve dependencies and run .NET code. So I guess that even some for runtime interceptors are already made, making the job easier to be done.

> Is it to avoid building other code in the same unit test project?
Many unit test projects use shared helpers to simplify individual unit tests. To figure out the helpers required, the majority process of building would be required.

Actually it was just one example of how powerful can runtime interceptors be in development. And about what you said, I think these are just a quick dependency scan can detect these codes too. Even if this scan takes too long, there could be a thing for programs to annouce relationships between classes and dependencies before build time, so even that simple scan is not needed anymore.

**Comment by En3Tho** [comment]
My understanding is that you want to have interpteter-based runtime shipping right with sdk and have that as an option to run your programs without a "build to cil" step?

**Comment by huoyaoyuan** [comment]
> And about what you said, I think these are just a quick dependency scan can detect these codes too. Even if this scan takes too long, there could be a thing for programs to annouce relationships between classes and dependencies before build time

"Scanning annotations" is really big work during compilation. Creating such annotations separately also requires heavy compiler or manual work.

This fits more about compiler optimization that reuses unchanged partial information from previous runs.

**Comment by Suchiman** [comment]
The first big problem that would come to my mind would be: C# does not use "import X from file", it has `using namespace` and types can be in arbitrary files that need not be named the same as the type that is contained within, so when you write out `SomeType.DoSomething()`, it would need to search all the files in the project and all referenced projects to find `SomeType` and ensure there's no conflicting type named `SomeType`. Sounds like you'd be most of the way to a compilation there at that point.

**Comment by huoyaoyuan** [comment]
Look at the two examples:
```csharp
// file1.cs
void M(SomeType x) => x.DoSomeThing();

// file2.cs
static void DoSomeThing(this object x) {}

// file3.cs
static void DoSomeThing(this SomeType x) {}
```
The existence of file3 changes overload resolution.

---
```csharp
// file1.cs
using Bar;
namespace Foo
{
    class C : SomeClass {}
}

// file2.cs
namespace Bar
{
    class SomeClass {}
}

// file3.cs
namespace Foo
{
    class SomeClass {}
}
```
The existence of file3 changes type name resolution. There's not a conflict since same namespace wins.

---
Anything else in the same compilation unit can affect the code behavior. C# isn't designed for single-pass compilation or execution. It doesn't work like script languages or C/C++.
